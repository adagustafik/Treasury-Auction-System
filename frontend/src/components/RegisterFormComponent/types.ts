export interface RegisterFormProp {
  page: PageType;
}

export type PageType = "register" | "profile";
