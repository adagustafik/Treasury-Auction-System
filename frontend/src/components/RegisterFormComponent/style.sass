@use '../../App'

.register-form
    @include App.centeredPosition($width: 45vw)

.register-form input
    @include App.inputFields
    font-weight: 700   

.register-form input::placeholder
    font-size: 1.8vh
    color: 80% black

.register-form button
    @include App.inputFields
    background-color: App.$buttonColor
    color: black
    font-size: 2.1vh
    font-weight: 500
    border: 1px solid App.$buttonColor c