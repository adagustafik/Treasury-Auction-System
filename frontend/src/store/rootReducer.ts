import { combineReducers } from "@reduxjs/toolkit";
import { auctionsReducer } from "./slices/auctions";
import { userReducer } from "./slices/user";
import { bidsReducer } from "./slices/bids";
import { purchasesReducer } from "./slices/purchases";

export const rootReducer = combineReducers({
  userReducer,
  auctionsReducer,
  purchasesReducer,
  bidsReducer,
});
