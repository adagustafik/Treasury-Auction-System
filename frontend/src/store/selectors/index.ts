export * from "./selectUser";
export * from "./selectAuctions";
export * from "./selectBids";
export * from "./selectPurchases";
