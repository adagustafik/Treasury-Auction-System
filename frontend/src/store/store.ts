import {
  configureStore,
  ThunkAction,
  Action,
  AnyAction,
  ThunkDispatch,
} from "@reduxjs/toolkit";
import thunkMiddleware from "redux-thunk";
import { rootReducer } from "./rootReducer";

export const store = configureStore({
  reducer: rootReducer,
  middleware: [thunkMiddleware],
});

export type AppDispatch = typeof store.dispatch &
  ThunkDispatch<RootState, null, AnyAction>;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
