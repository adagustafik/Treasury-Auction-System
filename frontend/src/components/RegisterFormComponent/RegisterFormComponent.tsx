import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import toast from "react-hot-toast";
import { fetchRegister } from "../../api";
import "./style.sass";
import { EMAIL_PATTERN, PASSWORD_PATTERN } from "../../configuration";
import { RegisterFormProp } from "./types";
import { selectUser } from "../../store/selectors";
import { useAppDispatch, useAppSelector } from "../../store/hooks";
import { fetchUpdateUser } from "../../api/updateUserRequest";
import { updateUserDataAction } from "../../store/slices/user";

export function RegisterFormComponent({ page }: RegisterFormProp) {
  const { email } = useAppSelector(selectUser);

  const [validated, setValidated] = useState(false);
  const [username, setUsername] = useState("");
  const emailInInit = page === "register" ? "" : String(email);
  const [emailIn, setEmailIn] = useState(emailInInit);
  const [password1, setPassword1] = useState("");
  const [password2, setPassword2] = useState("");

  const navigate = useNavigate();
  const dispatch = useAppDispatch();

  function validationOutsideOfForm(): boolean {
    if (page === "register" && username.length < 3) {
      toast("Please, provide valid username.");
      return false;
    }
    if (
      emailIn.length < 6 ||
      !emailIn.match(EMAIL_PATTERN) ||
      !password1.match(PASSWORD_PATTERN)
    ) {
      toast("Please, fill out all required fields.");
      return false;
    }
    if (password1 !== password2) {
      toast("Passwords do not match. Try again!");
      return false;
    }
    return true;
  }

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    setValidated(true);
    if (validationOutsideOfForm()) {
      if (page === "register") {
        const body = { username, email: emailIn, password: password1 };
        fetchRegister(body)
          .then((res) => {
            if (res.success) {
              navigate("/");
              toast(res.success);
            }
          })
          .catch((err) => {
            toast(err.message);
          });
      }
      if (page === "profile") {
        const body = { email: emailIn, password: password1 };
        fetchUpdateUser(body)
          .then((res) => {
            if (res.success) {
              dispatch(updateUserDataAction(emailIn));
              navigate("/dashboard");
              toast(res.success);
            }
          })
          .catch((err) => {
            toast(err.message);
          });
      }
    }
  };

  return (
    <div className="register-form">
      <Form noValidate validated={validated} onSubmit={handleSubmit}>
        {page === "register" && (
          <Form.Control
            required
            minLength={3}
            type="text"
            placeholder="username"
            value={username}
            onChange={(event: React.ChangeEvent<HTMLInputElement>) =>
              setUsername(event.target.value)
            }
          />
        )}
        <Form.Control.Feedback type="invalid">
          Username has to be at least 3 characters long
        </Form.Control.Feedback>
        <Form.Control
          required
          minLength={6}
          pattern={EMAIL_PATTERN}
          type="text"
          placeholder={email}
          value={emailIn}
          onChange={(event: React.ChangeEvent<HTMLInputElement>) =>
            setEmailIn(event.target.value)
          }
        />
        <Form.Control.Feedback type="invalid">
          Please, provide valid email address
        </Form.Control.Feedback>
        <Form.Control
          required
          pattern={PASSWORD_PATTERN}
          type="password"
          placeholder="password"
          value={password1}
          onChange={(event: React.ChangeEvent<HTMLInputElement>) =>
            setPassword1(event.target.value)
          }
        />
        <Form.Control.Feedback type="invalid">
          Password must be at least 8 characters long and contain a number,
          special character and capital letter
        </Form.Control.Feedback>
        <Form.Control
          required
          pattern={PASSWORD_PATTERN}
          type="password"
          placeholder="confirm your password"
          value={password2}
          onChange={(event: React.ChangeEvent<HTMLInputElement>) =>
            setPassword2(event.target.value)
          }
        />
        <Form.Control.Feedback type="invalid">
          Password must be at least 8 characters long and contain a number,
          special character and capital letter
        </Form.Control.Feedback>
        {page === "register" && <Button type="submit">REGISTER</Button>}
        {page === "profile" && <Button type="submit">UPDATE</Button>}
      </Form>
    </div>
  );
}
