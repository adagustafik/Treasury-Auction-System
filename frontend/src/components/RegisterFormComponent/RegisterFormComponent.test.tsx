import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import { store } from "../../store/store";
import { RegisterFormComponent } from "./RegisterFormComponent";

it("renders correctly for register page", () => {
  const registerForm = renderer
    .create(
      <Provider store={store}>
        <BrowserRouter>
          <RegisterFormComponent page="register" />
        </BrowserRouter>
      </Provider>
    )
    .toJSON();
  expect(registerForm).toMatchSnapshot();
});
