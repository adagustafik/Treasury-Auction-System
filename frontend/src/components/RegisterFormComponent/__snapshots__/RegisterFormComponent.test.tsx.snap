// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly for register page 1`] = `
<div
  className="register-form"
>
  <form
    className=""
    noValidate={true}
    onSubmit={[Function]}
  >
    <input
      className="form-control"
      minLength={3}
      onChange={[Function]}
      placeholder="username"
      required={true}
      type="text"
      value=""
    />
    <div
      className="invalid-feedback"
    >
      Username has to be at least 3 characters long
    </div>
    <input
      className="form-control"
      minLength={6}
      onChange={[Function]}
      pattern="(^(.+)@(.+)\\\\.(.{2,})$)"
      placeholder="email"
      required={true}
      type="text"
      value=""
    />
    <div
      className="invalid-feedback"
    >
      Please, provide valid email address
    </div>
    <input
      className="form-control"
      onChange={[Function]}
      pattern="(?=.*[0-9])(?=.*[A-Z])(?=.*[!-/:-@_{-~}]).{8,}"
      placeholder="password"
      required={true}
      type="password"
      value=""
    />
    <div
      className="invalid-feedback"
    >
      Password must be at least 8 characters long and contain a number, special character and capital letter
    </div>
    <input
      className="form-control"
      onChange={[Function]}
      pattern="(?=.*[0-9])(?=.*[A-Z])(?=.*[!-/:-@_{-~}]).{8,}"
      placeholder="confirm your password"
      required={true}
      type="password"
      value=""
    />
    <div
      className="invalid-feedback"
    >
      Password must be at least 8 characters long and contain a number, special character and capital letter
    </div>
    <button
      className="btn btn-primary"
      disabled={false}
      type="submit"
    >
      REGISTER
    </button>
  </form>
</div>
`;
